File: software/ios/Tucklet/Views/Screens.swift
// Screens.swift
// Home / Library / Transfer / Settings, plus the plain-language state badge.
// License: PolyForm Noncommercial 1.0.0

import SwiftUI

// MARK: - Plain-language state badge (never the word "cache")

struct StateBadge: View {
    let state: ItemState
    var body: some View {
        let (icon, text) = display
        Label(text, systemImage: icon)
            .font(.caption2)
            .padding(.horizontal, 6).padding(.vertical, 3)
            .background(Capsule().fill(Brand.accent.opacity(0.12)))
            .foregroundStyle(Brand.ink)
            .accessibilityLabel(text)
    }
    private var display: (String, String) {
        switch state {
        case .onPhone: return ("iphone", "On phone")
        case .onTucklet: return ("circle.hexagongrid", "On Tucklet")
        case .temporary(let exp):
            if let exp { return ("clock", "Temporary · \(Self.remaining(exp))") }
            return ("clock", "Temporary")
        }
    }
    private static func remaining(_ epoch: Int64) -> String {
        let secs = max(epoch - Int64(Date().timeIntervalSince1970), 0)
        return TransferEstimator.human(seconds: UInt32(min(secs, Int64(UInt32.max))))
    }
}

// MARK: - Home

struct HomeView: View {
    @EnvironmentObject var model: AppModel
    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: 20) {
                    presenceCard
                    if !model.lastOffloadIds.isEmpty { undoBanner }
                    summaryCard
                    quickActions
                }.padding()
            }
            .background(Brand.paper.ignoresSafeArea())
            .navigationTitle("Tucklet")
        }
    }

    private var undoBanner: some View {
        HStack {
            Image(systemName: "checkmark.circle.fill").foregroundStyle(Brand.accent)
            Text("Backed up \(model.lastOffloadIds.count) photos").foregroundStyle(Brand.ink)
            Spacer()
            Button("Undo") { Task { await model.undoLastOffload() } }
                .font(.subheadline.bold()).foregroundStyle(Brand.accent)
        }
        .padding().background(RoundedRectangle(cornerRadius: 16).fill(Brand.accent.opacity(0.08)))
    }

    private var presenceCard: some View {
        VStack(spacing: 8) {
            Image(systemName: model.connectionState == .connected ? "circle.hexagongrid.fill" : "circle.hexagongrid")
                .font(.system(size: 44)).foregroundStyle(Brand.accent)
            switch model.connectionState {
            case .connected:
                if let s = model.status {
                    Text("\(s.batteryPercent)%\(s.charging ? " · charging" : "")")
                        .font(.headline).foregroundStyle(Brand.ink)
                }
                Text(model.freeText).font(.subheadline).foregroundStyle(Brand.muted)
                Text(model.storageDescription).font(.caption).foregroundStyle(Brand.muted)
            case .connecting: ProgressView("Finding your Tucklet…")
            case .idle: Text("Bring your Tucklet near").foregroundStyle(Brand.muted)
            case .error(let e): Text(e).font(.caption).foregroundStyle(.red)
            }
        }
        .frame(maxWidth: .infinity).padding(24)
        .background(RoundedRectangle(cornerRadius: 20).fill(.white))
    }

    private var summaryCard: some View {
        let pending = model.pendingBackupCount
        return HStack {
            Image(systemName: pending == 0 ? "checkmark.seal.fill" : "tray.and.arrow.up")
                .foregroundStyle(Brand.accent)
            Text(pending == 0 ? "Everything's backed up" : "\(pending) photos waiting to back up")
                .foregroundStyle(Brand.ink)
            Spacer()
        }
        .padding().background(RoundedRectangle(cornerRadius: 16).fill(.white))
    }

    private var quickActions: some View {
        VStack(spacing: 12) {
            NavigationLink { LibraryView() } label: {
                actionRow("Free up space", "tray.and.arrow.up", "Move photos to your Tucklet")
            }
            NavigationLink { LibraryView() } label: {
                actionRow("Get photos", "tray.and.arrow.down", "Bring a copy back to your phone")
            }
        }
    }
    private func actionRow(_ title: String, _ icon: String, _ sub: String) -> some View {
        HStack(spacing: 14) {
            Image(systemName: icon).font(.title3).foregroundStyle(Brand.accent).frame(width: 28)
            VStack(alignment: .leading) {
                Text(title).font(.headline).foregroundStyle(Brand.ink)
                Text(sub).font(.caption).foregroundStyle(Brand.muted)
            }
            Spacer(); Image(systemName: "chevron.right").foregroundStyle(Brand.muted)
        }
        .padding().background(RoundedRectangle(cornerRadius: 16).fill(.white))
    }
}

// MARK: - Library (per-app organization; metadata/thumbnails only)

struct LibraryView: View {
    @EnvironmentObject var model: AppModel
    @State private var selection = Set<String>()
    @State private var showTransfer = false
    @State private var kind: TransferKind = .offload

    var grouped: [(app: String, items: [MediaItem])] {
        model.libraryGroups()
    }

    var body: some View {
        NavigationStack {
            List {
                if model.pendingBackupCount > 0 {
                    Section {
                        Button {
                            // Quick action: select all pending + offload.
                            selection = Set(pendingItems.map(\.id)); kind = .offload; showTransfer = true
                        } label: {
                            Label("Back up \(model.pendingBackupCount) new photos", systemImage: "tray.and.arrow.up")
                                .foregroundStyle(Brand.accent)
                        }
                    }
                }
                ForEach(grouped, id: \.app) { group in
                    Section(group.app) {
                        ForEach(group.items) { item in row(item) }
                    }
                }
            }
            .background(Brand.paper)
            .navigationTitle("Library")
            .toolbar {
                if !selection.isEmpty {
                    ToolbarItemGroup(placement: .bottomBar) {
                        if selectionHasOnPhone {
                            Button { kind = .offload; showTransfer = true } label: { Label("Free up space", systemImage: "tray.and.arrow.up") }
                        }
                        Spacer()
                        if selectionHasOnTucklet {
                            Button { kind = .load; showTransfer = true } label: { Label("Get a copy", systemImage: "tray.and.arrow.down") }
                        }
                    }
                }
            }
            .task {
                if model.manifest == nil { await model.loadLibrary() }
                if model.onPhoneItems.isEmpty { await model.refreshPhoneLibrary() }
            }
            .sheet(isPresented: $showTransfer) {
                TransferSheet(kind: kind, items: selectedItems) {
                    selection.removeAll()
                }
            }
        }
    }

    private var allItems: [MediaItem] { model.onPhoneItems + (model.manifest?.items ?? []) }
    private var selectedItems: [MediaItem] { allItems.filter { selection.contains($0.id) } }
    private var pendingItems: [MediaItem] {
        let onTucklet = Set((model.manifest?.items ?? []).map { $0.id })
        return model.onPhoneItems.filter { !onTucklet.contains($0.id) }
    }
    private var selectionHasOnPhone: Bool {
        selectedItems.contains { if case .onPhone = $0.state { return true } else { return false } }
    }
    private var selectionHasOnTucklet: Bool {
        selectedItems.contains { if case .onTucklet = $0.state { return true } else { return false } }
    }

    private func row(_ item: MediaItem) -> some View {
        HStack(spacing: 12) {
            Button {
                if selection.contains(item.id) { selection.remove(item.id) } else { selection.insert(item.id) }
            } label: {
                Image(systemName: selection.contains(item.id) ? "checkmark.circle.fill" : "circle")
                    .foregroundStyle(selection.contains(item.id) ? Brand.accent : Brand.muted)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(selection.contains(item.id) ? "Selected" : "Not selected")

            NavigationLink {
                ItemDetailView(item: item)
            } label: {
                HStack(spacing: 12) {
                    Thumbnail(item: item).frame(width: 44, height: 44).clipShape(RoundedRectangle(cornerRadius: 8))
                    VStack(alignment: .leading, spacing: 4) {
                        Text(item.name).foregroundStyle(Brand.ink).lineLimit(1)
                        HStack { StateBadge(state: item.state); Text(ByteFormat.short(item.sizeBytes)).font(.caption2).foregroundStyle(Brand.muted) }
                    }
                    Spacer()
                }
            }
            .accessibilityLabel("\(item.name), \(item.state.label)")
        }
    }
}

/// Async thumbnail that works for both on-phone (PhotoKit) and on-Tucklet (HTTP).
struct Thumbnail: View {
    @EnvironmentObject var model: AppModel
    let item: MediaItem
    @State private var image: UIImage?
    var body: some View {
        ZStack {
            if let image {
                Image(uiImage: image).resizable().scaledToFill()
            } else {
                RoundedRectangle(cornerRadius: 8).fill(Brand.accent.opacity(0.10))
                Image(systemName: item.isVideo ? "video" : "photo").foregroundStyle(Brand.muted)
            }
        }
        .task(id: item.id) { image = await model.thumbnail(for: item) }
    }
}

/// Detail for one item: large thumbnail, state, origin, and the right actions.
struct ItemDetailView: View {
    @EnvironmentObject var model: AppModel
    @Environment(\.dismiss) var dismiss
    let item: MediaItem
    @State private var busy = false

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                Thumbnail(item: item).frame(height: 240).clipShape(RoundedRectangle(cornerRadius: 16))
                Text(item.name).font(.headline).foregroundStyle(Brand.ink)
                StateBadge(state: item.state)
                VStack(alignment: .leading, spacing: 6) {
                    detail("From", item.origin.app)
                    if let album = item.origin.album { detail("Album", album) }
                    detail("Size", ByteFormat.short(item.sizeBytes))
                    detail("Type", item.isVideo ? "Video" : "Photo")
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding().background(RoundedRectangle(cornerRadius: 14).fill(.white))

                actions
            }.padding()
        }
        .background(Brand.paper.ignoresSafeArea())
        .navigationTitle("Details").navigationBarTitleDisplayMode(.inline)
    }

    @ViewBuilder private var actions: some View {
        switch item.state {
        case .onTucklet:
            Button { run { try await model.restore(item) } } label: {
                actionLabel("Put back on phone", "tray.and.arrow.down")
            }
            Button(role: .destructive) { Task { busy = true; await model.delete(item); busy = false; dismiss() } } label: {
                actionLabel("Delete from Tucklet", "trash")
            }
        case .onPhone:
            Button { run { try await model.offloadItem(item); await model.finishOffload(deletedIds: []) } } label: {
                actionLabel("Back up to Tucklet", "tray.and.arrow.up")
            }
        case .temporary:
            Text("This is a temporary copy on your phone.").foregroundStyle(Brand.muted)
        }
    }

    private func run(_ work: @escaping () async throws -> Void) {
        Task { busy = true; try? await work(); busy = false; dismiss() }
    }
    private func detail(_ k: String, _ v: String) -> some View {
        HStack { Text(k).foregroundStyle(Brand.muted); Spacer(); Text(v).foregroundStyle(Brand.ink) }
    }
    private func actionLabel(_ t: String, _ icon: String) -> some View {
        Label(t, systemImage: icon).font(.headline).frame(maxWidth: .infinity).padding()
            .background(Brand.accent.opacity(busy ? 0.4 : 1)).foregroundStyle(.white)
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

// MARK: - Transfer sheet (the most important screen: shows the time up front)

struct TransferSheet: View {
    @EnvironmentObject var model: AppModel
    @Environment(\.dismiss) var dismiss
    let kind: TransferKind
    let items: [MediaItem]
    var onDone: () -> Void = {}

    @State private var mode: TransferMode = .batch
    @State private var policy: TemporaryPolicy = .oneWeek
    @StateObject private var engine = TransferEngineHolder()

    private var transferItems: [TransferItem] {
        items.map { TransferItem(id: $0.id, sizeBytes: $0.sizeBytes, mime: $0.mime) }
    }
    private var estimate: TransferEstimate {
        TransferEstimator.estimate(sizes: items.map(\.sizeBytes), link: model.currentLinkProfile())
    }

    var body: some View {
        NavigationStack {
            VStack(spacing: 22) {
                VStack(spacing: 6) {
                    Text(kind == .offload ? "Free up space" : "Get a copy").font(.title2.bold()).foregroundStyle(Brand.ink)
                    Text("\(items.count) items · \(ByteFormat.short(estimate.bytesTotal))").foregroundStyle(Brand.muted)
                }
                // The headline number: estimated time, before anything moves.
                VStack(spacing: 4) {
                    Text("About \(estimate.human)").font(.system(size: 34, weight: .bold)).foregroundStyle(Brand.accent)
                    Text("over \(transportLabel)").font(.caption).foregroundStyle(Brand.muted)
                }
                .padding().frame(maxWidth: .infinity)
                .background(RoundedRectangle(cornerRadius: 18).fill(.white))

                Picker("When", selection: $mode) {
                    Text("Now").tag(TransferMode.batch)
                    Text("Automatically").tag(TransferMode.trickle)
                }.pickerStyle(.segmented)

                if kind == .load {
                    HStack {
                        Text("Keep on phone for").foregroundStyle(Brand.ink); Spacer()
                        Picker("", selection: $policy) {
                            ForEach(TemporaryPolicy.allCases, id: \.self) { Text($0.label).tag($0) }
                        }.tint(Brand.accent)
                    }.padding().background(RoundedRectangle(cornerRadius: 14).fill(.white))
                }

                if let p = engine.engine?.progress, engine.engine?.isRunning == true {
                    VStack(spacing: 8) {
                        ProgressView(value: Double(p.bytesDone), total: Double(max(p.bytesTotal, 1)))
                        Text("\(p.itemsDone) of \(p.itemsTotal) · \(TransferEstimator.human(seconds: p.etaSeconds)) left")
                            .font(.caption).foregroundStyle(Brand.muted)
                    }
                }

                Spacer()
                Button {
                    Task { await start() }
                } label: {
                    Text(mode == .batch ? "Start" : "Turn on auto backup")
                        .font(.headline).frame(maxWidth: .infinity).padding()
                        .background(Brand.accent).foregroundStyle(.white)
                        .clipShape(RoundedRectangle(cornerRadius: 14))
                }
            }
            .padding()
            .background(Brand.paper.ignoresSafeArea())
            .navigationBarTitleDisplayMode(.inline)
            .toolbar { ToolbarItem(placement: .cancellationAction) { Button("Close") { dismiss() } } }
            .onAppear { policy = model.defaultTemporaryPolicy }
        }
    }

    private var transportLabel: String {
        switch model.capabilities?.bestWireless {
        case .wifiAware: return "Wi-Fi"
        case .softAp: return "Wi-Fi"
        default: return "Wi-Fi"
        }
    }

    private func start() async {
        guard let e = model.makeTransferEngine() else { return }
        engine.engine = e
        let req = TransferRequest(kind: kind, mode: mode, items: transferItems,
                                  temporaryPolicy: kind == .load ? policy : nil)

        if mode == .trickle {
            // Auto mode: just enable trickle and let the scheduler drip in the
            // background; nothing to run synchronously here.
            model.trickleEnabled = true
            onDone(); dismiss(); return
        }

        switch kind {
        case .offload:
            let completed = await e.runBatch(req) { item in
                try await model.offloadItem(item)
            }
            // Free up space: after a successful upload, remove the originals
            // from the phone and record them for Undo.
            await model.finishOffload(deletedIds: completed)
        case .load:
            let chosen = policy
            _ = await e.runBatch(req) { item in
                try await model.loadItem(item, policy: chosen)
            }
        }

        if e.lastError == nil { onDone(); dismiss() }
    }
}

/// Small holder so the sheet can own an engine created after model wiring.
@MainActor final class TransferEngineHolder: ObservableObject {
    @Published var engine: TransferEngine?
}

// MARK: - Settings

struct SettingsView: View {
    @EnvironmentObject var model: AppModel
    @State private var showForget = false
    var body: some View {
        NavigationStack {
            Form {
                Section("Backup") {
                    Toggle("Back up automatically", isOn: $model.trickleEnabled)
                    Toggle("Only while charging", isOn: $model.trickleOnlyWhileCharging)
                        .disabled(!model.trickleEnabled)
                }
                Section("Temporary copies") {
                    Picker("Default keep time", selection: $model.defaultTemporaryPolicy) {
                        ForEach(TemporaryPolicy.allCases, id: \.self) { Text($0.label).tag($0) }
                    }
                }
                Section("Device") {
                    LabeledContent("Storage", value: model.storageDescription)
                    if let s = model.status {
                        LabeledContent("Battery", value: "\(s.batteryPercent)%")
                        LabeledContent("Firmware", value: s.firmwareVersion)
                    }
                }
                if !model.lastOffloadIds.isEmpty {
                    Section("Recent") {
                        Button {
                            Task { await model.undoLastOffload() }
                        } label: {
                            Label("Undo last backup (\(model.lastOffloadIds.count) photos)", systemImage: "arrow.uturn.backward")
                        }
                    }
                }
                Section("Paired phones") {
                    Text("This phone").foregroundStyle(Brand.ink)
                    Button("Forget this Tucklet", role: .destructive) { showForget = true }
                } footer: {
                    Text("Forgetting stops this phone from connecting. To also erase this phone from the Tucklet itself, hold the button on the device for 5 seconds to factory-reset it.")
                }
            }
            .navigationTitle("Settings")
            .confirmationDialog("Forget this Tucklet on this phone?", isPresented: $showForget, titleVisibility: .visible) {
                Button("Forget", role: .destructive) { Task { await model.forgetTucklet() } }
                Button("Cancel", role: .cancel) {}
            }
        }
    }
}

File: software/android/app/src/main/java/app/tucklet/ui/Screens.kt
// Screens.kt
// The Compose UI: onboarding gate, Home, Library (unified on-phone + on-Tucklet
// with real thumbnails), item detail + restore, transfer sheet (estimate up
// front), and Settings. Plain-language states throughout (UX_SPEC).
//
// License: PolyForm Noncommercial 1.0.0
package app.tucklet.ui

import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.tucklet.core.TransferEstimator
import app.tucklet.protocol.*
import app.tucklet.store.AppViewModel

@Composable
fun TuckletApp(vm: AppViewModel, onStartPairing: () -> Unit) {
    val paired by vm.isPaired.collectAsState()
    if (!paired) {
        OnboardingScreen(vm, onStartPairing)
    } else {
        MainScaffold(vm)
    }
}

@Composable
private fun OnboardingScreen(vm: AppViewModel, onStartPairing: () -> Unit) {
    Column(
        Modifier.fillMaxSize().background(Brand.paper).padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Icon(Icons.Filled.Favorite, null, tint = Brand.accent, modifier = Modifier.size(64.dp))
        Spacer(Modifier.height(16.dp))
        Text("Meet your Tucklet", fontSize = 26.sp, fontWeight = FontWeight.Bold, color = Brand.ink)
        Spacer(Modifier.height(8.dp))
        Text(
            "Bring it close and tap to connect. You'll only do this once — after that it just works whenever it's near.",
            color = Brand.muted,
        )
        Spacer(Modifier.height(32.dp))
        Button(onClick = onStartPairing, modifier = Modifier.fillMaxWidth()) {
            Text("Connect my Tucklet")
        }
        val err by vm.errorText.collectAsState()
        err?.let { Spacer(Modifier.height(12.dp)); Text(it, color = MaterialTheme.colorScheme.error) }
    }
}

private enum class Tab(val label: String, val icon: androidx.compose.ui.graphics.vector.ImageVector) {
    HOME("Home", Icons.Filled.Home),
    LIBRARY("Library", Icons.Filled.GridView),
    SETTINGS("Settings", Icons.Filled.Settings),
}

@Composable
private fun MainScaffold(vm: AppViewModel) {
    var tab by remember { mutableStateOf(Tab.HOME) }
    var detail by remember { mutableStateOf<MediaItem?>(null) }
    var transfer by remember { mutableStateOf<Pair<TransferKind, List<MediaItem>>?>(null) }

    LaunchedEffect(Unit) { vm.connect(); vm.loadLibrary() }

    Scaffold(
        containerColor = Brand.paper,
        bottomBar = {
            NavigationBar(containerColor = androidx.compose.ui.graphics.Color.White) {
                Tab.entries.forEach { t ->
                    NavigationBarItem(
                        selected = tab == t,
                        onClick = { tab = t },
                        icon = { Icon(t.icon, t.label) },
                        label = { Text(t.label) },
                    )
                }
            }
        }
    ) { pad ->
        Box(Modifier.padding(pad)) {
            when (tab) {
                Tab.HOME -> HomeScreen(vm)
                Tab.LIBRARY -> LibraryScreen(vm, onOpen = { detail = it }, onTransfer = { k, items -> transfer = k to items })
                Tab.SETTINGS -> SettingsScreen(vm)
            }
        }
    }

    detail?.let { item ->
        ItemDetailSheet(vm, item, onClose = { detail = null })
    }
    transfer?.let { (kind, items) ->
        TransferSheet(vm, kind, items, onClose = { transfer = null })
    }
}

@Composable
private fun HomeScreen(vm: AppViewModel) {
    val status by vm.status.collectAsState()
    val conn by vm.connection.collectAsState()
    val lastOffload by vm.lastOffloadIds.collectAsState()
    Column(Modifier.fillMaxSize().padding(16.dp), verticalArrangement = Arrangement.spacedBy(16.dp)) {
        Card { Column(Modifier.padding(20.dp).fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) {
            Icon(Icons.Filled.Album, null, tint = Brand.accent, modifier = Modifier.size(44.dp))
            Spacer(Modifier.height(8.dp))
            val s = status
            if (s != null) {
                Text("${s.batteryPercent}%${if (s.charging) " · charging" else ""}", fontWeight = FontWeight.Bold, color = Brand.ink)
                Text("${vm.byteFormat(s.freeBytes)} free of ${vm.byteFormat(s.totalBytes)}", color = Brand.muted)
            } else if (conn == app.tucklet.store.AppRepository.Conn.CONNECTING) {
                Text("Finding your Tucklet…", color = Brand.muted)
            } else {
                Text("Bring your Tucklet near", color = Brand.muted)
            }
        } }
        if (lastOffload.isNotEmpty()) {
            Card(colors = CardDefaults.cardColors(containerColor = Brand.accent.copy(alpha = 0.08f))) {
                Row(Modifier.padding(16.dp).fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Icon(Icons.Filled.CheckCircle, null, tint = Brand.accent)
                    Spacer(Modifier.width(8.dp))
                    Text("Backed up ${lastOffload.size} photos", color = Brand.ink)
                    Spacer(Modifier.weight(1f))
                    TextButton(onClick = { vm.undoLastOffload() }) { Text("Undo") }
                }
            }
        }
        val pending = vm.pendingBackupCount()
        Card { Row(Modifier.padding(16.dp).fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Icon(if (pending == 0) Icons.Filled.Verified else Icons.Filled.Upload, null, tint = Brand.accent)
            Spacer(Modifier.width(8.dp))
            Text(if (pending == 0) "Everything's backed up" else "$pending photos waiting to back up", color = Brand.ink)
        } }
    }
}

@Composable
private fun LibraryScreen(
    vm: AppViewModel,
    onOpen: (MediaItem) -> Unit,
    onTransfer: (TransferKind, List<MediaItem>) -> Unit,
) {
    val manifest by vm.manifest.collectAsState()
    val onPhone by vm.onPhoneItems.collectAsState()
    val selection = remember { mutableStateListOf<String>() }
    val groups = remember(manifest, onPhone) { vm.libraryGroups() }

    Column(Modifier.fillMaxSize()) {
        if (selection.isNotEmpty()) {
            Row(Modifier.fillMaxWidth().padding(12.dp), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                val selected = groups.flatMap { it.second }.filter { selection.contains(it.id) }
                if (selected.any { it.itemState is ItemState.OnPhone })
                    Button(onClick = { onTransfer(TransferKind.OFFLOAD, selected); selection.clear() }) { Text("Free up space") }
                if (selected.any { it.itemState is ItemState.OnTucklet })
                    OutlinedButton(onClick = { onTransfer(TransferKind.LOAD, selected); selection.clear() }) { Text("Get a copy") }
            }
        }
        LazyColumn(Modifier.fillMaxSize()) {
            groups.forEach { (app, items) ->
                item { Text(app, fontWeight = FontWeight.Bold, color = Brand.muted, modifier = Modifier.padding(16.dp, 12.dp, 16.dp, 4.dp)) }
                items(items, key = { it.id }) { item ->
                    LibraryRow(vm, item, selected = selection.contains(item.id),
                        onToggle = { if (selection.contains(item.id)) selection.remove(item.id) else selection.add(item.id) },
                        onOpen = { onOpen(item) })
                }
            }
        }
    }
}

@Composable
private fun LibraryRow(
    vm: AppViewModel, item: MediaItem, selected: Boolean,
    onToggle: () -> Unit, onOpen: () -> Unit,
) {
    Row(Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 6.dp), verticalAlignment = Alignment.CenterVertically) {
        IconButton(onClick = onToggle) {
            Icon(if (selected) Icons.Filled.CheckCircle else Icons.Filled.RadioButtonUnchecked, "select",
                tint = if (selected) Brand.accent else Brand.muted)
        }
        Thumb(vm, item, 44)
        Spacer(Modifier.width(12.dp))
        Column(Modifier.weight(1f).clickable { onOpen() }) {
            Text(item.name, color = Brand.ink, maxLines = 1)
            Row(verticalAlignment = Alignment.CenterVertically) {
                StateChip(item.itemState)
                Spacer(Modifier.width(8.dp))
                Text(vm.byteFormat(item.sizeBytes), fontSize = 12.sp, color = Brand.muted)
            }
        }
        Icon(Icons.Filled.ChevronRight, null, tint = Brand.muted)
    }
}

@Composable
private fun Thumb(vm: AppViewModel, item: MediaItem, size: Int) {
    var bmp by remember(item.id) { mutableStateOf<Bitmap?>(null) }
    LaunchedEffect(item.id) { bmp = vm.thumbnail(item) }
    Box(Modifier.size(size.dp).clip(RoundedCornerShape(8.dp)).background(Brand.accent.copy(alpha = 0.10f)),
        contentAlignment = Alignment.Center) {
        val b = bmp
        if (b != null) Image(b.asImageBitmap(), null, modifier = Modifier.fillMaxSize())
        else Icon(if (item.isVideo) Icons.Filled.Videocam else Icons.Filled.Photo, null, tint = Brand.muted)
    }
}

@Composable
private fun StateChip(state: ItemState) {
    val (icon, label) = when (state) {
        is ItemState.OnPhone -> Icons.Filled.Smartphone to "On phone"
        is ItemState.OnTucklet -> Icons.Filled.Album to "On Tucklet"
        is ItemState.Temporary -> Icons.Filled.Schedule to "Temporary"
    }
    Row(verticalAlignment = Alignment.CenterVertically) {
        Icon(icon, null, tint = Brand.accent, modifier = Modifier.size(14.dp))
        Spacer(Modifier.width(4.dp))
        Text(label, fontSize = 12.sp, color = Brand.muted)
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ItemDetailSheet(vm: AppViewModel, item: MediaItem, onClose: () -> Unit) {
    ModalBottomSheet(onDismissRequest = onClose, containerColor = Brand.paper) {
        Column(Modifier.padding(20.dp).fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) {
            Thumb(vm, item, 200)
            Spacer(Modifier.height(12.dp))
            Text(item.name, fontWeight = FontWeight.Bold, color = Brand.ink)
            StateChip(item.itemState)
            Spacer(Modifier.height(8.dp))
            Text("From ${item.origin.app}", color = Brand.muted)
            Spacer(Modifier.height(16.dp))
            when (item.itemState) {
                is ItemState.OnTucklet -> {
                    Button(onClick = { vm.restore(item); onClose() }, modifier = Modifier.fillMaxWidth()) { Text("Put back on phone") }
                    Spacer(Modifier.height(8.dp))
                    OutlinedButton(onClick = { vm.delete(item); onClose() }, modifier = Modifier.fillMaxWidth()) { Text("Delete from Tucklet") }
                }
                is ItemState.OnPhone ->
                    Button(onClick = { vm.runOffload(listOf(item)) { onClose() } }, modifier = Modifier.fillMaxWidth()) { Text("Back up to Tucklet") }
                is ItemState.Temporary -> Text("This is a temporary copy on your phone.", color = Brand.muted)
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TransferSheet(vm: AppViewModel, kind: TransferKind, items: List<MediaItem>, onClose: () -> Unit) {
    var policy by remember { mutableStateOf(TemporaryPolicy.ONE_WEEK) }
    val est = remember(items) { vm.estimate(items) }
    ModalBottomSheet(onDismissRequest = onClose, containerColor = Brand.paper) {
        Column(Modifier.padding(20.dp).fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) {
            Text(if (kind == TransferKind.OFFLOAD) "Free up space" else "Get a copy",
                fontSize = 22.sp, fontWeight = FontWeight.Bold, color = Brand.ink)
            Text("${items.size} items · ${vm.byteFormat(est.bytesTotal)}", color = Brand.muted)
            Spacer(Modifier.height(16.dp))
            Card(Modifier.fillMaxWidth()) { Column(Modifier.padding(20.dp), horizontalAlignment = Alignment.CenterHorizontally) {
                Text("About ${est.human}", fontSize = 32.sp, fontWeight = FontWeight.Bold, color = Brand.accent)
                Text("over Wi-Fi", fontSize = 12.sp, color = Brand.muted)
            } }
            if (kind == TransferKind.LOAD) {
                Spacer(Modifier.height(12.dp))
                Text("Keep on phone for", color = Brand.ink)
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    TemporaryPolicy.entries.forEach { p ->
                        FilterChip(selected = policy == p, onClick = { policy = p }, label = { Text(p.label) })
                    }
                }
            }
            Spacer(Modifier.height(20.dp))
            Button(
                onClick = {
                    if (kind == TransferKind.OFFLOAD) vm.runOffload(items) { onClose() }
                    else vm.runLoad(items, policy) { onClose() }
                },
                modifier = Modifier.fillMaxWidth(),
            ) { Text("Start") }
            Spacer(Modifier.height(8.dp))
        }
    }
}

@Composable
private fun SettingsScreen(vm: AppViewModel) {
    val context = LocalContext.current
    var auto by remember { mutableStateOf(true) }
    val status by vm.status.collectAsState()
    val caps by vm.capabilities.collectAsState()
    val lastOffload by vm.lastOffloadIds.collectAsState()

    LazyColumn(Modifier.fillMaxSize().padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        item {
            Card { Column(Modifier.padding(16.dp)) {
                Text("Backup", fontWeight = FontWeight.Bold, color = Brand.ink)
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Text("Back up automatically", Modifier.weight(1f), color = Brand.ink)
                    Switch(checked = auto, onCheckedChange = {
                        auto = it
                        if (it) vm.enableTrickle(context) else vm.disableTrickle(context)
                    })
                }
            } }
        }
        item {
            Card { Column(Modifier.padding(16.dp)) {
                Text("Device", fontWeight = FontWeight.Bold, color = Brand.ink)
                Text("Storage: ${caps?.storage?.label ?: "—"}", color = Brand.muted)
                status?.let {
                    Text("Battery: ${it.batteryPercent}%", color = Brand.muted)
                    Text("Firmware: ${it.firmwareVersion}", color = Brand.muted)
                }
            } }
        }
        if (lastOffload.isNotEmpty()) {
            item { OutlinedButton(onClick = { vm.undoLastOffload() }, modifier = Modifier.fillMaxWidth()) {
                Text("Undo last backup (${lastOffload.size} photos)")
            } }
        }
        item {
            Card { Column(Modifier.padding(16.dp)) {
                Text("Paired phones", fontWeight = FontWeight.Bold, color = Brand.ink)
                Text("This phone", color = Brand.ink)
                Spacer(Modifier.height(8.dp))
                TextButton(onClick = { vm.forget() }) { Text("Forget this Tucklet", color = MaterialTheme.colorScheme.error) }
                Text(
                    "Forgetting stops this phone from connecting. To also erase this phone from the Tucklet itself, hold its button for 5 seconds to factory-reset it.",
                    fontSize = 12.sp, color = Brand.muted,
                )
            } }
        }
    }
}
